src/device-manager/java/src/nl/Weave/DeviceManager/TestMain.java

```diff
@@ -661,6 +661,88 @@ void testWdmClientDataSinkSetBigIntegerFlushRefreshGetBigInteger(WeaveDeviceMana
         System.out.println("testWdmClientDataSinkSetBigIntegerFlushRefreshGetBigInteger Succeeded");
     }
 
+    void testWdmClientGetEvents(WeaveDeviceManager deviceMgr)
+    {
+        WdmClientFactory wdmClientFactory = new WdmClientFactory();
+        WdmClient mockWdmClient = wdmClientFactory.create(deviceMgr);
+
+        mockWdmClient.setCompletionHandler(this);
+
+        GenericTraitUpdatableDataSink localSettingsTrait;
+
+        ResourceIdentifier resourceIdentifier = new ResourceIdentifier();
+        localSettingsTrait = mockWdmClient.newDataSink(resourceIdentifier, 20, 0, "/");
+        localSettingsTrait.setCompletionHandler(this);
+
+        mockWdmClient.setNodeId(new BigInteger("-2"));
+        mockWdmClient.setEventFetchingTimeout(1);
+
+        mockWdmClient.beginRefreshDataWithEvents();
+        ExpectSuccess("RefreshData"); // Need to wait the data, or the events are not prepared
+        String event = mockWdmClient.getEvents();
+
+        // Just check if it can return data, the content is verified in happy tests
+        // "[]".length() = 2
+        if (event.length() <= 2)
+            throw new TestFailedException("testWdmClientGetEvents");
+
+        try { Thread.sleep(5000); }
+        catch (Exception ex) { }
+
+        mockWdmClient.beginRefreshDataWithEvents();
+        ExpectSuccess("RefreshData"); // Need to wait the data, or the events are not prepared
+        event = mockWdmClient.getEvents();
+
+        // Just check if it can return data, the content is verified in happy tests
+        // "[]".length() = 2
+        if (event.length() <= 2)
+            throw new TestFailedException("testWdmClientGetEvents");
+
+        try { Thread.sleep(5000); }
```

Comment:
Why is there a sleep() call here?

Reply:
Added a comment to explain the sleep -- for the mock-device generating more events.

Reply:
Is there no way to have the mock client emit events deterministically? Ideally a test shouldn't have to delay 5 seconds for a result (and instead would simulate the passing of 5 seconds).

Reply:
the device could send multiple of notification for the long time if there is huge amounts of events, therefore we add timeout here, and mobile client would receive at least one notification or multiple notification until timeout.
here 5000 is used for test so that device can have enough time to generate events.